e2e_playwright/st_dialog_test.py

```diff
@@ -0,0 +1,183 @@
+# Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022-2024)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from playwright.sync_api import Page, expect
+
+from e2e_playwright.conftest import ImageCompareFunction, wait_for_app_run
+
+modal_test_id = "stModal"
+
+
+def open_dialog_with_images(app: Page):
+    app.get_by_text("Open Dialog with Images").click()
+
+
+def open_dialog_without_images(app: Page):
+    app.get_by_text("Open Dialog without Images").click()
+
+
+def open_nondismissible_dialog(app: Page):
+    app.get_by_text("Open non-dismissible Dialog").click()
+
+
+def open_largewidth_dialog(app: Page):
+    app.get_by_text("Open large-width Dialog").click()
+
+
+def open_sidebar_dialog(app: Page):
+    app.get_by_text("Open Sidebar-Dialog").click()
+
+
+def click_to_dismiss(app: Page):
+    # Click somewhere outside the close popover container:
+    app.get_by_test_id("stModal").click(position={"x": 0, "y": 0})
+
+
+def test_displays_dialog_properly(app: Page):
+    """Test that dialog is displayed properly."""
+    open_dialog_with_images(app)
+    wait_for_app_run(app)
+    main_dialog = app.get_by_test_id(modal_test_id)
+    expect(main_dialog).to_have_count(1)
+
+
+def test_dialog_closes_properly(app: Page):
+    """Test that dialog closes after clicking on action button."""
+    open_dialog_with_images(app)
+    wait_for_app_run(app)
+    main_dialog = app.get_by_test_id(modal_test_id)
+    expect(main_dialog).to_have_count(1)
+    close_button = main_dialog.get_by_test_id("stButton").locator("button").first
+    close_button.scroll_into_view_if_needed()
+    close_button.click()
+    wait_for_app_run(app)
+    main_dialog = app.get_by_test_id(modal_test_id)
+    expect(main_dialog).to_have_count(0)
+
+
+def test_dialog_dismisses_properly(app: Page):
+    """Test that dialog is dismissed properly after clicking on modal close (= dismiss)."""
+    open_dialog_with_images(app)
+    wait_for_app_run(app)
+    main_dialog = app.get_by_test_id(modal_test_id)
+    expect(main_dialog).to_have_count(1)
+
+    click_to_dismiss(app)
+    expect(main_dialog).not_to_be_visible()
+
+    wait_for_app_run(app)
```

Comment:
Is dismissing the dialog also triggering a rerun?

Reply:
it does not! I have removed this wait and clarified this in the dialog's docstring